### DISCUSSION TITLE
SX1276 in FSK mode

### DISCUSSION THREAD

**Discussion by mrquincle** [jgromes/RadioLib#393] [General]
Dear all, and Jan in particular,

I'm using the SX1276 (Uno + Dragino shield v1.4) in FSK mode to try to synthesize an 2-FSK signal for a device in the SDR band, a so-called "social alarm", see https://github.com/mrquincle/social_alarm for the spectrum analysis and corresponding GNU radio blocks.

Now I've browsing this repository and the discussion quite extensively, but I haven't been able to make the radio do what I want. I first started with trying to receive data, but that's harder to debug, hence I focused on transmitting.

* I see that if `preambleLength` is 0, there's no output, as soon as I set it to 8, there's output on the SDR. I've a sync word set as well. Is it fine to have no preamble set?
* It seems that if I enable OOK directly with `beginFSK` I get an error, while I can enable it later just fine. 
* If Manchester encoding is chosen, will the sync words be Manchester encoded as well? 
* Can preamble polarity, 0x55 / 0xAA be set?
* It seems that the `transmit` command always timeouts. What could be the reason?
* The timeouts I see are 5 seconds.

If I look at the code in `SX127x.cpp`, the timeout here seems to be 5000000 microseconds, hence 5000 ms, while the comment suggests it should be 5 ms. On the air I also see 5 s rather than 5 ms.

```
  } else if(modem == SX127X_FSK_OOK) {
    // calculate timeout (5ms + 500 % of expected time-on-air)
    uint32_t timeout = 5000000 + (uint32_t)((((float)(len * 8)) / (_br * 1000.0)) * 5000000.0);
``` 

However, I see this in more transmit functions so I'm assuming I'm not understanding it right for now.

In direct mode, I can see a single spectral line when I'm transmitting a tone on pin 4 (so I know I've the right pin). If I transmit another tone, nothing changes. I suppose that if I set the radio to say 868.25 MHz, a tone of 1 kHz will be at 868.251 MHz and a tone of 10 kHz will be at 868.26 MHz, correct?

These are - I know - general questions, I will over time add .ino files and how the spectrum looks like bit by bit. Answers will be already highly appreciated.

**Comment by jgromes** [comment]
To answer your questions:

> Is it fine to have no preamble set?

The datasheet doesn't seem to forbid this, but I wouldn't be surprised if it was actually not supported - especially since the SX127x register for preamble configuration is in byte units, so calling `setPreambleLength(8);` will actually write `1` to the register, it seems reasonable to assume nothing smaller is supported.

> It seems that if I enable OOK directly with beginFSK I get an error

That's strange - what is the error?

> If Manchester encoding is chosen, will the sync words be Manchester encoded as well?

The payload, CRC, packet length and address are encoded; the sync word is not.

> Can preamble polarity, 0x55 / 0xAA be set?

SX1276 supports this, but it's not implemented in RadioLib. Currently 0x55 preamble is used.

Regarding the transmit timeouts, I'll check with the sketch in your repository (https://github.com/mrquincle/social_alarm/blob/main/dragino/dragino_2fsk.ino).

> In direct mode, I can see a single spectral line when I'm transmitting a tone on pin 4 (so I know I've the right pin)

You will see a single spectral line regardless of the pin connection - that's just the radio sending an unmodulated carrier. Is your pin 4 connected to DIO2? On most off-the-shelf SX1276 modules, only DIO0 and DIO1 are connected.

**Reply by mrquincle** [reply]
I will commit a new sketch in which I transmit to single this behaviour out (because in the mentioned sketch I don't transmit). However, just from the code you would expect `5000000` μs to mean 5 seconds, correct?

**Reply by jgromes** [reply]
> It seems that if I enable OOK directly with beginFSK I get an error

I was not able to replicate this issue, using the parameters you have in your sketch, the module initialzies correctly (I'm using a different frequency because I don't have SX1276, only SX1278).

```c++
radio.beginFSK(434.0, 1.96, 25.0, 50, 10, 24, true);
```

> It seems that the transmit command always timeouts

I did not see any timeouts transmitting a test "Hello World!" packet using your configuration (0xAA sync word, Manchester encoding etc.). IIRC, the 5000000 us (5 sec) timeout value was intentional change since 5 ms was too short for low datarates, I probably just forgot to change the comment.

**Reply by mrquincle** [reply]
I've made a very simple sketch at https://github.com/mrquincle/social_alarm/blob/main/dragino/fsk_transmit/fsk_transmit.ino and am in this way using the defaults w.r.t. preamble length, crc, sync word, variable packet length mode, etc.

# Code

```
int pin_cs    = 10;
int pin_reset =  9;
int pin_dio0  =  2;
int pin_dio1  =  6;
int pin_dio2  =  7;
int pin_dio5  =  8;

// Global radio object
SX1276 radio = new Module(pin_cs, pin_dio0, pin_reset, pin_dio1);

void setup() {
  radio.beginFSK(868.0, 1.96, 25.0, 50, 10, 24, true);
}

void loop() {
  sprintf(outbuf, "Msgno=%d", cnt++);
  Serial.print("Msg");
  Serial.println(cnt++);
  int state = radio.transmit(outbuf);
  if (state != ERR_NONE) {
    Serial.print(F("Unable to transmit data, code "));
    Serial.println(state);
    return;
  }
  delay(2000);
}
```
The pin layout is just from the [dragino v1.4 datasheet](https://github.com/mrquincle/social_alarm/blob/main/datasheets/Lora%20Shield%20v1.4.sch.pdf).

# Verbose output

I've added the listing I get setting VERBOSITY of the RadioLib at the end of that file. I'll repeat here only the last part:

```

13:57:41.346 -> Send message of length 7
13:57:41.346 -> R 1 29  
13:57:41.346 -> R 1 29  
13:57:41.346 -> W 1 29  
13:57:41.346 -> R 1 29  
13:57:41.346 -> R 1 29  
13:57:41.346 -> R 40  0 
13:57:41.346 -> W 40  0 
13:57:41.346 -> R 40  0 
13:57:41.346 -> R 1 29  
13:57:41.346 -> W 3E  FF  
13:57:41.346 -> W 3F  FF  
13:57:41.346 -> W 0 7 
13:57:41.346 -> R 30  90  
13:57:41.346 -> W 0 4D  73  67  6E  6F  3D  30  
13:57:41.346 -> R 1 29  
13:57:41.346 -> R 1 29  
13:57:41.346 -> W 1 2B  
13:57:41.346 -> R 1 2B  
13:57:46.518 -> R 1 2B  
13:57:46.518 -> W 3E  FF  
13:57:46.518 -> W 3F  FF  
13:57:46.518 -> R 1 2B  
13:57:46.518 -> R 1 2B  
13:57:46.518 -> W 1 29  
13:57:46.518 -> R 1 2B  
13:57:46.518 -> R 1 2B  
13:57:46.518 -> R 1 29  
13:57:46.518 -> Unable to transmit data, code -5
```

If I'm correct, for transmitting in packet mode, I only need dio0 and dio1 (apart from cs). If I read the following quote, this looks like what is happening on this radio. 

> "The  transmission  of  packet  data  is  initiated  by  the  Packet  Handler  only  if  the  chip  is  in  Tx  mode  and  the  transmission  
> condition defined by TxStartCondition is fulfilled. If transmission condition is not fulfilled then the packet handler transmits a 
> preamble sequence until the condition is met. This happens only if the preamble length /= 0, otherwise it transmits a zero or 
> one until the condition is met to transmit the packet data." 

# DIO5
If I check table 30 in the [SX1276 datasheet](https://raw.githubusercontent.com/mrquincle/social_alarm/main/datasheets/sx1276_77_78_79.pdf) for packet mode, there's some functionality assigned to pin_dio5 as well.

```
DIO5:
00: ClkOut
01: PllLock
10: Data
11: ModeReady
```

Regarding the fact that there's no pin to indicate dio5 in the module constructor, is this all redundant and is the FIFO sufficient?

# TX

If I print (on the first `SPItransfer` in `Module.cpp`) the values of the pins, then _rx is set to 6 (dio1), and _tx is set to 0xFF (not connected). I assume I shouldn't take the names of these pins too literal and consider them just fifo (in or out). Correct?

# Hardware

Next step I will take a look with the logic, but I have to say I would be very surprised if the Dragino is actually **not** connected as indicated in the datasheets. That's the entire purpose of a shield.

**Reply by jgromes** [reply]
I tried your sketch and got a -107 error - turns out there was an issue in OOK mode configuration from the `begin` method, that was fixed in the latest commit. However, After fixing it, I did not get any timeouts.

> If I'm correct, for transmitting in packet mode, I only need dio0 and dio1 (apart from cs).

Actually, for transmitting packets, only DIO0 is needed (to detect timeouts). From the quote (which I assume is from datasheet), it looks like preamble length set to zero will prevent all transmissions.

> If I check table 30 in the SX1276 datasheet for packet mode, there's some functionality assigned to pin_dio5 as well.

DIO5 is not used in the library.

> I assume I shouldn't take the names of these pins too literal and consider them just fifo

Those are for modules that use UART instead of SPI. One of them is also used additional GPIO for SPI modules (for example, second IRQ on SX127x or BUSY pin on SX126x).

**Reply by mrquincle** [reply]
Check, https://github.com/jgromes/RadioLib/commit/6475507274f2d1b1cddec6759a9cfe69409654b3 looks indeed like the fix for the error I got regarding OOK mode configuration.

Calling configFSK() in https://github.com/jgromes/RadioLib/commit/9824ac32b6311afb07fa94e12037f6a03b623789 looks like something that might fix my timeouts.

Check... check... Ah, it's not a release yet. Pulling from git and voila, yes it works!!

![image](https://user-images.githubusercontent.com/1428585/139212635-9c971baa-baa6-4b6f-aafd-30b1124e0594.png)

Super duper nice! Thanks for fixing things so fast. I know how hard it can be if you don't have exactly the same hardware! Much appreciated!

**Reply by jgromes** [reply]
You're welcome - I just published a new release, since there was a lot of new stuff added since 4.5.0, so if you're using Arduino IDE, the library manager should pick it up in couple of hours.

**Comment by mrquincle** [comment]
# Direct Mode
Regarding direct mode, ahhh. Dumb! By connecting the DIO2 to pin 4, I can indeed see the modulations. So, that's one thing out of the way. :+1: Here I generate tones of 0.5 kHz, 5 kHz, 10 kHz, 15 kHz, 25 kHz and 50 kHz at 869.265 MHz. At 50 kHz you see the tones that are most widely spaced apart (at 869.215 MHz and 869.315 MHz). The others correspondingly, sweet.

![image](https://user-images.githubusercontent.com/1428585/139040349-eee0e457-cb65-497d-be7f-23e9e3c838cc.png)

Now, let's see if I did something else dumb in FSK mode as well. :-)

**Reply by jgromes** [reply]
Glad you managed to resolve it - out of curiosity, what SDR software are you using? I'm not familiar with the one on the screenshot.

**Reply by mrquincle** [reply]
This is SigDigger, you can find it at https://github.com/BatchDrake/SigDigger and it's available as AppImage, so it's easy to try out. I really like its visuals.
